utilities/bump_versions.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import copy
+import re
+
+version_regex = re.compile(r'(\d+)\.(\d+)\.(\d+)(-\w+)?(-\w+)?')
+
+class Version:
+    def __init__(self, version_str):
+        match = version_regex.match(version_str)
+        major = int(match.group(1))
+        minor = int(match.group(2))
+        patch = int(match.group(3))
+        qualifier1 = match.group(4)
+        qualifier2 = match.group(5)
+        variant = ''
+        snapshot = False
+        if qualifier1 and qualifier2:
+            if qualifier2 == '-SNAPSHOT':
+                variant = qualifier1
+                snapshot = True
+            else:
+                variant = qualifier1 + qualifier2
+        elif qualifier1:
+            if qualifier1 == '-SNAPSHOT':
+                snapshot = True
+            else:
+                variant = qualifier1
+        self.major = major
+        self.minor = minor
+        self.patch = patch
+        self.variant = variant
+        self.snapshot = snapshot
+
+    def bump(self, bump_type):
+        if bump_type == 'minor':
+            self.bump_minor()
+        elif bump_type == 'patch':
+            self.bump_patch()
+        else:
+            raise ValueError('invalid bump_type: ' + bump_type)
+
+    def bump_minor(self):
+        self.minor += 1
+        self.patch = 0
+
+    def bump_patch(self):
+        self.patch += 1
+
+    def set_snapshot(self, snapshot):
+        self.snapshot = snapshot
+
+    def __str__(self):
+        mmp = str(self.major) + '.' + str(self.minor) + '.' + str(self.patch)
```

Comment:
please use 
```python
"%s%s" % (var1, var2)
```
for string concatenation. This applies to all other places where `+` is used.

Reply:
I'll use the newer idiom `'{}'.format(x)`